torch/optim/optimizer.py

```diff
@@ -909,6 +925,8 @@ def update_group(
             group: Dict[str, Any], new_group: Dict[str, Any]
         ) -> Dict[str, Any]:
             new_group["params"] = group["params"]
+            if 'param_names' in group and 'param_names' not in new_group:
+                new_group["param_names"] = group["param_names"]
```

Comment:
Wait I might be confused--but doesn't this contradict the earlier statement about overriding? This looks like the new_group will only get written IF not previously defined? aka it will not be overridden?

Reply:
Note that new_group is the saved_group (from the loaded state_dict).

if `param_names` in both group and new_group: keep (and set) the content of the `new_group["param_names"]` (override previous content). Thus, no changes are needed.
if `param_names` in new_group only: keep the content of the `new_group["param_names"]` (also no changes are needed)
if `param_names` in `group` **only** (this case, see code below): copy the content of `group["param_names"]` to `new_group["param_names"]` which will be set to the optimizer to preserve the names.

In this case, e.g.:
``` python
optimizer_named = optim_cls(model2.named_parameters(), **optim_input.kwargs)
optimizer_named .load_state_dict(state_dict_without_names)
```

This behavior is important for addressing situations where you have a `non_named_state_dict` and want to obtain a `named_optimizer` or `named_state_dict`. This is achieved by loading the `non_named_state_dict` into a named optimizer, and when saving its `state_dict`, the parameter names will be included.

Please let me know if I've missed anything.

Reply:
The new_group being from the saved state_dict is a great distinction. thanks for clarifying